onnxruntime/core/providers/cpu/math/einsum_auxiliary_ops.cc

```diff
@@ -0,0 +1,296 @@
+    int64_t first_dim_axis = -1;  // This is the axis eventually occupied by the first_dim
+
+    // If one of the diagonal dimensions is one of the 2 innermost dims, then leave it as such
+    // so as to avoid transpose overhead
+    if (first_dim == rank - 2) { // If rank - 2 is occupied by first_dim, keep it there
+      permutation[rank - 2] = first_dim;
+      first_dim_axis = rank - 2;
+    } else {
+      if (second_dim != rank - 2) {  // If rank - 2 is not occupied by second_dim, then put first_dim there
+        permutation[rank - 2] = first_dim;
+        first_dim_axis = rank - 2;
+      } else {  // If rank - 2 is occupied by second_dim, then put first_dim in rank - 1
+        permutation[rank - 1] = first_dim;
+        first_dim_axis = rank - 1;
+        preserve_innermost_dim_val = true; // We always want to preserve the dim value of the first_dim
+      }
+    }
+
+    // Put the second_dim in the dim not occupied by the first_dim
+    if (first_dim_axis != rank - 1) {
+      permutation[rank - 1] = second_dim;
+    } else {
+      permutation[rank - 2] = second_dim;
+    }
+
+    int64_t iter = 0;
+    for (int64_t i = 0; i < rank; ++i) {
+      if (i != first_dim && i != second_dim) {
+        permutation[iter++] = i;
+      }
+    }
+    ORT_ENFORCE(iter == rank - 2);
+
+    // Permuatate the input so that the dims from which we need the diagonal forms the innermost dims
+    auto transposed = Transpose(input, permutation, allocator);
+
+    // Parse the diagonal from the innermost dims
+    output = DiagonalInnermostDims(transposed, preserve_innermost_dim_val, allocator);
+
+    // Swap back the dimensions to the original axes ordering using a "reverse permutation"
+
+    // Find the "reverse" permutation
+    iter = 0;
+    std::vector<size_t> reverse_permutation(rank, 0);
+    for (const auto& perm : permutation) {
+      reverse_permutation[perm] = iter++;
+    }
+
+    // Permutate using the reverse permutation to get back the original axes ordering
+    output = Transpose(output, reverse_permutation, allocator);
+  } else {
+    // No transposing required
+    output = DiagonalInnermostDims(input, preserve_innermost_dim_val, allocator);
+  }
+
+  // Make copy of the transposed output
+  auto output_dims = output.Shape().GetDims();
+
+  // Unsqueeze the reduced dim
+  auto iter = output_dims.begin() + second_dim;
+  output_dims.erase(iter);
+
+  // Reshape to output_dims
+  CreateReshapedView(output, output_dims);
+
+  return output;
+}
+
+// Explicit template instantiation
+
+// float
+template Tensor MatMul<float>(const Tensor& input_1, const Tensor& input_2, 
+    AllocatorPtr allocator, concurrency::ThreadPool* tp);
+template Tensor ReduceSum<float>(const Tensor& input, const std::vector<int64_t>& reduce_axes, 
+    AllocatorPtr allocator, concurrency::ThreadPool* tp);
+template Tensor ReduceSum<float>(const Tensor& input, int64_t axis, AllocatorPtr allocator, concurrency::ThreadPool* tp);
+
```

Comment:
nit: doesn't look like clang-format has run correctly on this file as the args would usually be aligned inside the opening '('

Reply:
Ran them through clang again manually just in case.